src/imitation/data/dataset.py

```diff
@@ -0,0 +1,168 @@
+        """In-memory dict data sampler that samples in epoch-order.
+
+        No sample from `data_map` can be returned an X+1th time by `sample()` until
+        every other sample has been returned X times.
+
+        Args:
+            data_map: A mapping from keys to np.ndarray values, where every value has
+                an equal number of rows. This argument is not mutated, as each array
+                is immediately copied.
+            shuffle: If true, then shuffle the dataset upon initialization and at the
+                end of every epoch.
+        """
+        super().__init__(data_map)
+        self._next_idx = 0
+        self._shuffle = shuffle
+        if shuffle:
+            self.shuffle_dataset()
+
+    def shuffle_dataset(self) -> None:
+        """Shuffles the data_map in place."""
+        perm = np.arange(self.size())
+        np.random.shuffle(perm)
+        for key in self.data_map:
+            self.data_map[key] = self.data_map[key][perm]
+
+    def sample(self, n_samples: int) -> Dict[str, np.ndarray]:
+        if n_samples <= 0:
+            raise ValueError(f"n_samples={n_samples} must be positive.")
+        samples_accum = {k: [] for k in self.data_map.keys()}
+        n_samples_remain = n_samples
+        while n_samples_remain > 0:
+            n_samples_actual, sample = self._sample_bounded(n_samples_remain)
+            n_samples_remain -= n_samples_actual
+            for k, v in sample.items():
+                samples_accum[k].append(v)
+
+        result = {k: np.concatenate(v) for k, v in samples_accum.items()}
+        assert all(len(v) == n_samples for v in result.values())
+        return result
+
+    def _sample_bounded(
+        self, n_samples_request: int
+    ) -> Tuple[int, Dict[str, np.ndarray]]:
+        """Like `.sample()`, but allowed to return fewer samples on epoch boundaries."""
+        assert n_samples_request > 0
+        if self._next_idx >= self.size():
+            self._next_idx = 0
+            if self._shuffle:
+                self.shuffle_dataset()
+
+        n_samples_actual = min(n_samples_request, self.size() - self._next_idx)
+        assert n_samples_actual > 0
+
+        result = {}
+        for key in self.data_map:
+            result[key] = self.data_map[key][
+                self._next_idx : self._next_idx + n_samples_actual
+            ]
+            assert len(result[key]) == n_samples_actual
+        self._next_idx += n_samples_actual
+        return n_samples_actual, result
+
+
+class RandomDictDataset(EpochOrderDictDataset):
+    """In-memory data sampler that uniformly samples with replacement."""
+
+    def sample(self, n_samples: int) -> Dict[str, np.ndarray]:
+        if n_samples <= 0:
+            raise ValueError(f"n_samples={n_samples} must be positive.")
+        inds = np.random.randint(self.size(), size=n_samples)
+        return {k: v[inds] for k, v in self.data_map.items()}
+
+
+class TransitionsDictDatasetAdaptor(Dataset[types.Transitions]):
```

Comment:
`S = TypeVar('S', bound=types.Transitions)` was actually my first attempt, but I ran into a ton of infuriating `pytype` errors that I couldn't cleanly work around, so I gave up and circled back to a totally generic type.

For example, after instantiating `ds = TransistionsDictDatasetAdaptor(transitions_with_rew)`, pytype would think that `ds` is of type `TransisitionsDictDatasetAdaptor[Any]` (even though the [PEP specification](https://www.python.org/dev/peps/pep-0484/#instantiating-generic-classes-and-type-erasure) says it should be `...[TransitionsWithRew]`) leading to an error down the line with Adversarial, which requires `expert_dataset: Dataset[Transitions]`.

Reply:
Turns out drawing 3 shuffled batches of 10000 samples by indexing a permuted view from a dataset of 10000 is about 10x slower than doing the same thing without a view.

```
----------------------------------------------------------------------------------------------------- benchmark: 8 tests ----------------------------------------------------------------------------------------------------
Name (time in us)                                    Min                   Max                  Mean            StdDev                Median               IQR            Outliers          OPS            Rounds  Iterations
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_shuffle_benchmark[10000-True-False]        136.2401 (3.92)       137.5046 (3.94)       136.8203 (3.93)     0.4704 (11.71)      136.8517 (3.94)     0.5964 (11.21)         2;0   7,308.8578 (0.25)          5       10000
test_shuffle_benchmark[10000-True-True]       1,208.6027 (34.78)    1,225.0590 (35.13)    1,213.1699 (34.86)    6.7838 (168.96)   1,211.2302 (34.83)    6.1334 (115.29)        1;1     824.2868 (0.03)          5        1000
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean

```

Reply:
Oh wild. I guess this is because random memory reads are a lot slower than reading contiguous chunks both for caching and RAM latency reasons. Would not have predicted this, glad you benchmarked it!